// preq/index-new.js
"use strict";

var P = require('bluebird');

function wrapCreateSocket(agent) {
    var createSocket = agent.createSocket;
    agent.createSocket = function(req, options) {
        var s = createSocket.call(agent, req, options);
        // Set up a connect timeout if connectTimeout option is set
        if (options.connectTimeout && !s.connectTimeoutTimer) {
            // Set up a timer
            s.connectTimeoutTimer = setTimeout(function () {
                var e = new Error("ETIMEDOUT");
                e.code = "ETIMEDOUT";
                s.end();
                s.emit("error", e);
                s.emit('close');
                //console.error('[AgentConnectTimeout ' + hostHeader + ':' + options.port + ']');
            }, options.connectTimeout);
            s.once('connect',  function() {
                if (this.connectTimeoutTimer) {
                    //console.log('canceled timeout');
                    clearTimeout(s.connectTimeoutTimer);
                    this.connectTimeoutTimer = undefined;
                }
            });
        }
        return s;
    };
}

var spdy = require('spdy');
var httpAgent = spdy.createAgent({
    connectTimeout: 60000,
    // Setting this too high (especially 'Infinity') leads to high
    // (hundreds of mb) memory usage in the agent under sustained request
    // workloads. 250 should be a reasonable upper bound for practical
    // applications.
    maxSockets: 250,

    // Optional SPDY options
    spdy: {
        plain: true,
        ssl: true,
        //version: 3 // Force SPDY version
    }
});
wrapCreateSocket(httpAgent);

require('http').globalAgent = httpAgent;
require('https').globalAgent = httpAgent;

var util = require('util');

var request = P.promisify(require('request'));

function getOptions(uri, o, method) {
    if (!o || o.constructor !== Object) {
        if (uri) {
            if (typeof uri === 'object') {
                o = uri;
            } else {
                o = { uri: uri };
            }
        } else {
            throw new Error('preq options missing!');
        }
    } else {
        o.uri = uri;
    }
    o.method = method;
    if (o.body && o.body instanceof Object) {
        if (o.headers && /^application\/json/.test(o.headers['content-type'])) {
            o.body = JSON.stringify(o.body);
        } else if (o.method === 'post') {
            o.form = o.body;
            o.body = undefined;
        }
    }

    if ((o.method === 'get' || o.method === 'put')
            && o.retries === undefined) {
        // Idempotent methods: Retry once by default
        o.retries = 1;
    }

    if (o.query) {
        o.qs = o.query;
        o.query = undefined;
    }

    // Set a timeout by default
    if (o.timeout === undefined) {
        o.timeout = 2 * 60 * 1000; // 2 minutes
    }

    // Default to binary requests (return buffer)
    if (!o.encoding) {
        o.encoding = null;
    }

    // Default pool options: Don't limit the number of sockets
    if (!o.pool) {
        o.pool = {
            maxSockets: Infinity,
            http: httpAgent,
            https: httpAgent
        };
    }

    if (o.gzip === undefined && o.method === 'get') {
        o.gzip = true;
    }
    return o;
}

/*
 * Error instance wrapping HTTP error responses
 *
 * Has the same properties as the original response.
 */
function HTTPError(response) {
    Error.call(this);
    Error.captureStackTrace(this, HTTPError);
    this.name = this.constructor.name;
    this.message = response.status.toString();
    if (response.body && response.body.type) {
        this.message += ': ' + response.body.type;
    }

    for (var key in response) {
        this[key] = response[key];
    }
}
util.inherits(HTTPError, Error);


/*
 * Encapsulate the state associated with a single HTTP request
 */
function Request (method, url, options) {
    this.options = getOptions(url, options, method);
    this.retries = this.options.retries;
    this.timeout = this.options.timeout;
    this.retryDelay = this.options.retryDelay || 0; // start with no delay
}

Request.prototype.retry = function (err) {
    if (this.retries) {
        this.retries--;
        // grow the timeout exponentially, plus some fuzz
        this.timeout *= 3 + Math.random();

        return P.bind(this)
        .delay(this.retryDelay)
        .then(this.run);
    } else {
        throw err;
    }
};

Request.prototype.run = function () {
    var self = this;
    return P.try(request, this.options)
    .bind(this)
    .then(function(responses) {
        if (!responses || responses.length < 2) {
            return this.retry(new HTTPError({
                status: 502,
                body: {
                    type: 'empty_response',
                }
            }));
        } else {
            var response = responses[0];
            var body = responses[1]; // decompressed


            if (body && response.headers) {
                var contentType = response.headers['content-type'];
                if (/^text\/|application\/json\b/.test(contentType)) {
                    // Convert buffer to string
                    body = body.toString();
                }

                if (/^application\/(?:problem\+)?json\b/.test(contentType)) {
                    body = JSON.parse(body);
                }
            }

            var res = {
                status: response.statusCode,
                headers: response.headers,
                body: body
            };

            if (res.status >= 400) {
                if (res.status === 503
                        && /^[0-9]+$/.test(response.headers['retry-after'])
                        && parseInt(response.headers['retry-after']) * 1000 < self.options.timeout) {
                    self.retryDelay = parseInt(response.headers['retry-after']) * 1000;
                    return self.retry(new HTTPError(res));
                }

                throw new HTTPError(res);
            } else {
                return res;
            }
        }
    },
    function (err) {
        return this.retry(new HTTPError({
            status: 504,
            body: {
                type: 'internal_http_error',
                description: err.toString(),
                error: err,
                stack: err.stack,
                uri: self.options.uri,
                method: self.options.method,
            },
            stack: err.stack
        }));
    });
};

var preq = function preq (url, options) {
    var method = (options || url || {}).method || 'get';
    return new Request(method, url, options).run();
};

var methods = ['get','head','put','post','delete','trace','options','mkcol','patch'];
methods.forEach(function(method) {
    preq[method] = function (url, options) {
        return new Request(method, url, options).run();
    };
});

module.exports = preq;
